src/core/tools/contextValidator.ts

```diff
@@ -0,0 +1,170 @@
+import { Task } from "../task/Task"
+import { readLines } from "../../integrations/misc/read-lines"
+import { getModelMaxOutputTokens } from "../../shared/api"
+
+/**
+ * More aggressive buffer percentage specifically for file reading validation.
+ * This is separate from the global TOKEN_BUFFER_PERCENTAGE to provide extra safety
+ * when reading files without affecting other context window calculations.
+ */
+const FILE_READ_BUFFER_PERCENTAGE = 0.25 // 25% buffer for file reads
+
+export interface ContextValidationResult {
+	shouldLimit: boolean
+	safeMaxLines: number
+	reason?: string
+}
+
+/**
+ * Validates if a file can be safely read based on its size and current runtime context state.
+ * Reads lines incrementally and counts tokens as it goes, stopping when reaching the token limit.
+ * Returns a safe maxReadFileLine value to prevent context overflow.
+ */
+export async function validateFileSizeForContext(
+	filePath: string,
+	totalLines: number,
+	currentMaxReadFileLine: number,
+	cline: Task,
+): Promise<ContextValidationResult> {
+	try {
+		// Get actual runtime state from the task
+		const modelInfo = cline.api.getModel().info
+		const { contextTokens: currentContextTokens } = cline.getTokenUsage()
+		const contextWindow = modelInfo.contextWindow
+
+		// Get the model-specific max output tokens using the same logic as sliding window
+		const modelId = cline.api.getModel().id
+		const apiProvider = cline.apiConfiguration.apiProvider
+		const settings = await cline.providerRef.deref()?.getState()
+
+		// Map apiProvider to the format expected by getModelMaxOutputTokens
```

Comment:
Is there an existing helper that will produce a `format` given an `apiProvider`? This doesn't look future proof as new providers needs to get mapped to formats. I think it needs a utility function or it needs to be handled by `getModelMaxOutputTokens` directly.